src/core/net/dhcp6_server.cpp

```diff
@@ -0,0 +1,544 @@
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements DHCPv6 Server.
+ */
+#include <openthread-types.h>
+#include <common/code_utils.hpp>
+#include <common/encoding.hpp>
+#include <common/logging.hpp>
+#include <net/dhcp6_server.hpp>
+#include <thread/mle.hpp>
+#include <thread/thread_netif.hpp>
+
+using Thread::Encoding::BigEndian::HostSwap16;
+using Thread::Encoding::BigEndian::HostSwap32;
+
+namespace Thread {
+namespace Dhcp6 {
+
+Dhcp6Server::Dhcp6Server(ThreadNetif &aThreadNetif):
+    mSocket(aThreadNetif.GetIp6().mUdp),
+    mNetworkDataLocal(aThreadNetif.GetNetworkDataLocal()),
+    mMac(aThreadNetif.GetMac()),
+    mNetif(aThreadNetif)
+{
+    for (uint8_t i = 0; i < DHCP_PREFIX_MAX; i++)
+    {
+        memset(&mPrefixes[i], 0, sizeof(PrefixBuffer));
+    }
+
+    for (uint8_t i = 0; i < (DHCP_PREFIX_MAX - 1); i++)
+    {
+        mPrefixes[i].SetNext(&mPrefixes[i + 1]);
+    }
+
+    for (uint8_t i = 0; i < DHCP_ALLOW_ADDRESSES; i++)
+    {
+        memset(&mAddresses[i], 0, sizeof(AddressBuffer));
+    }
+
+    for (uint8_t i = 0; i < (DHCP_ALLOW_ADDRESSES - 1); i++)
+    {
+        mAddresses[i].SetNext(&mAddresses[i + 1]);
+    }
+
+    mPrefixesHead = NULL;
+    mPrefixesAvail = &mPrefixes[0];
+    mAddressesAvail = &mAddresses[0];
+}
+
+ThreadError Dhcp6Server::AddPrefix(otIp6Prefix &aIp6Prefix)
+{
+    ThreadError error = kThreadError_None;
+    PrefixBuffer *prefixBuffer = NULL;
+
+    VerifyOrExit(mPrefixesAvail, error = kThreadError_NoBufs);
+
+    prefixBuffer = mPrefixesAvail;
+    mPrefixesAvail = mPrefixesAvail->GetNext();
+
+    prefixBuffer->SetPrefix(aIp6Prefix);
+    prefixBuffer->SetAddressBuffer(NULL);
+
+    prefixBuffer->SetNext(mPrefixesHead);
+    mPrefixesHead = prefixBuffer;
+
+exit:
+    return error;
+}
+
+ThreadError Dhcp6Server::ReleasePrefix(otIp6Address *aIp6Address)
+{
+    ThreadError error = kThreadError_NotFound;
+    PrefixBuffer *prevPrefixBuffer = NULL;
+    PrefixBuffer *prefixBuffer = NULL;
+
+    VerifyOrExit(mPrefixesHead, ;);
+
+    prefixBuffer = mPrefixesHead;
+
+    while (prefixBuffer)
+    {
+        if (otIp6PrefixMatch(&(prefixBuffer->GetPrefix().mPrefix), aIp6Address) >= prefixBuffer->GetPrefix().mLength)
+        {
+            if (!prevPrefixBuffer)
+            {
+                mPrefixesHead = prefixBuffer->GetNext();
+            }
+            else
+            {
+                prevPrefixBuffer->SetNext(prefixBuffer->GetNext());
+            }
+
+            prefixBuffer->SetNext(mPrefixesAvail);
+            mPrefixesAvail = prefixBuffer;
+            ExitNow(error = kThreadError_None);
+        }
+
+        prevPrefixBuffer = prefixBuffer;
+        prefixBuffer = prefixBuffer->GetNext();
+    }
+
+exit:
+    return error;
+}
+
+bool Dhcp6Server::IsDhcpAgent()
+{
+    return ((mPrefixesHead) ? true : false);
+}
+
+bool Dhcp6Server::IsPrefixSingleton(const uint8_t *aIp6Address)
```

Comment:
"Singleton" means a set that contains a single element.  In this case, you have a method that checks if a set contains zero elements.  Maybe something like `PrefixHasClients()`.